File: python/pig-latin/pig_latin.py
import re

def translate_word(word):
    vowels = "aeiou"

    # Rule 1: Starts with vowel sound (a, e, i, o, u) or special cases 'xr', 'yt'
    if re.match(r'^[aeiou]|yt|xr', word):
        return word + "ay"

    # Find the consonant cluster
    # Match consonant(s), including 'y' if not first, potentially followed by 'qu'
    # Or handle 'y' as the first letter explicitly as a consonant
    if word.startswith('y'):
         # Match 'y' plus following consonants (if any)
         match = re.match(r'^(y[^aeiou]*)', word)
    else:
        # Match initial consonants (not 'y'), optionally followed by 'qu'
        # Or match initial consonants (not 'y')
        match = re.match(r'^([^aeiouy]*qu|[^aeiouy]+)', word)

    if match:
        consonant_cluster = match.group(0)
        # Check if the whole word matched (like 'rhythm' where 'y' is the vowel)
        if len(consonant_cluster) == len(word):
             # Check if 'y' is present and not the first letter (acts as vowel)
             y_pos = word.find('y', 1)
             if y_pos != -1:
                 # Split at 'y'
                 consonant_cluster = word[:y_pos]
                 remaining_word = word[y_pos:]
                 return remaining_word + consonant_cluster + "ay"
             else:
                 # No vowel 'y' found after consonants (e.g., "shh") - add "ay"
                 # Test cases don't cover this, but adding 'ay' seems consistent
                 return word + "ay"

        remaining_word = word[len(consonant_cluster):]
        return remaining_word + consonant_cluster + "ay"

    # Fallback (should not be reached with comprehensive regex, but for safety)
    return word + "ay"


def translate(text):
    words = text.split()
    translated_words = [translate_word(word) for word in words]
    return " ".join(translated_words)

File: javascript/pig-latin/pig-latin.js
//
// This is only a SKELETON file for the 'Pig Latin' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const VOWELS = /^[aeiou]|^(xr)|^(yt)/i; // Rule 1: Starts with vowel sound or xr, yt
const CONSONANT_QU = /^([^aeiou]*qu)(.*)/i; // Rule 3: Starts with consonants? + qu
const CONSONANT_Y = /^([^aeiou]+)(y.*)/i; // Rule 4: Starts with consonants + y (y acts as vowel)
const CONSONANT = /^([^aeiou]+)(.*)/i; // Rule 2: Starts with consonants

const translateWord = (word) => {
  if (VOWELS.test(word)) {
    return word + 'ay';
  }
  if (CONSONANT_QU.test(word)) {
    return word.replace(CONSONANT_QU, '$2$1ay');
  }
  // Rule 4 needs to be checked before Rule 2 for words like 'rhythm'
  // where 'y' acts as the first vowel sound after consonants.
  if (CONSONANT_Y.test(word)) {
    return word.replace(CONSONANT_Y, '$2$1ay');
  }
  if (CONSONANT.test(word)) {
    return word.replace(CONSONANT, '$2$1ay');
  }
  // Should not be reached for valid English words according to the rules
  return word;
};

export const translate = (phrase) => {
  if (!phrase) {
    return '';
  }
  return phrase.split(' ').map(translateWord).join(' ');
};
